Fixed passing common test modules path to unelevated Powershell

adityapatwardhan opened the pull request:
- Fixed the way common test modules are passed to elevated and unelevated powershell. Earlier, only elevated powershell got those through inheritance as a child process. Now we add them to the startup of the process.

- Fixed error reported by PSScriptAnalyzer about ? / Where-Object

- Converted all the parameters passed to powershell.exe to be a base64 encoded string to avoid complications with quotes.

- Removed code which was updated $env:PSModulePath as we do it in startup args for powershell process instead. 

- Added a way to disable -Quiet for Pester.

adityapatwardhan committed:
Fixed passing common test modules path to unelevated Powershell

adityapatwardhan committed:
Added way to disable -Quiet for Pester

SteveL-MSFT commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -479,35 +482,33 @@ function Invoke-OpenCover
-    $targetArgsElevated += @("-OutputFile $PesterLogElevated", "-OutputFormat $PesterLogFormat", "-Quiet")
-    $targetArgsUnelevated += @("-OutputFile $PesterLogUnelevated", "-OutputFormat $PesterLogFormat", "-Quiet")
+    $targetArgsElevated += @("-OutputFile $PesterLogElevated", "-OutputFormat $PesterLogFormat")
+    $targetArgsUnelevated += @("-OutputFile $PesterLogUnelevated", "-OutputFormat $PesterLogFormat")
```
Seems like there's lots of redundancy building the args, would it be better to have $targetArgs and then appending the unique parts of Elevated vs Unelevated closer to execution?

adityapatwardhan replied:
Yes, having a look if this can be simplified.

adityapatwardhan replied:
Removed some redundancy by extracting out a function.

SteveL-MSFT replied:
If `-OutputFormat` is the same, why not have it in $targetArgs instead?

SteveL-MSFT commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -479,35 +482,33 @@ function Invoke-OpenCover
-    $openCoverArgsElevated = "-target:$target","-register:user","-output:${outputLog}","-nodefaultfilters","-oldstyle","-hideskipped:all","-mergeoutput", "-filter:`"+[*]* -[Microsoft.PowerShell.PSReadLine]*`"", "-targetargs:`"$targetArgStringElevated`""
-    $openCoverArgsUnelevated = "-target:$target","-register:user","-output:${outputLog}","-nodefaultfilters","-oldstyle","-hideskipped:all", "-mergeoutput", "-filter:`"+[*]* -[Microsoft.PowerShell.PSReadLine]*`"", "-targetargs:`"$targetArgStringUnelevated`""
+    $openCoverArgsElevated = "-target:$target","-register:user","-output:${outputLog}","-nodefaultfilters","-oldstyle","-hideskipped:all","-mergeoutput", "-filter:`"+[*]* -[Microsoft.PowerShell.PSReadLine]*`"", "-targetargs:`"-EncodedCommand $base64targetArgsElevated`""
+    $openCoverArgsUnelevated = "-target:$target","-register:user","-output:${outputLog}","-nodefaultfilters","-oldstyle","-hideskipped:all", "-mergeoutput", "-filter:`"+[*]* -[Microsoft.PowerShell.PSReadLine]*`"", "-targetargs:`"-EncodedCommand $base64targetArgsUnelevated`""
```
Seems like there's lots of redundancy building the args, would it be better to have $targetArgs and then appending the unique parts of Elevated vs Unelevated closer to execution?

adityapatwardhan replied:
Yes, having a look if this can be simplified.

SteveL-MSFT replied:
I think the only difference here is `$base64targetArgs`.  You could have `$openCoverArgs` with all the common args, then on line 513, something like:

```powershell
& $OpenCoverBin ($openCoverArgsElevated + "-targetargs:`"-EncodedCommand $base64targetArgsElevated`"")
```

SteveL-MSFT commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -519,6 +520,9 @@ function Invoke-OpenCover
             # poll for process exit every 60 seconds
             # timeout of 6 hours
             $timeOut = ([datetime]::Now).AddHours(6)
```
How long does it typically take?  Not clear to me if 6 hours is a good value

adityapatwardhan replied:
It typically takes around 2.5 - 3 hours total, out of which maybe 2 hours for the unelevated one. '6 hours' was completely pulled out of thin air, to be substantially large enough.

SteveL-MSFT replied:
I think it would be worth putting that data in the comment to give context to 6 hours

adityapatwardhan replied:
Fixed.

SteveL-MSFT commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -465,8 +466,10 @@ function Invoke-OpenCover
+    $updatedEnvPath = "${PowerShellExeDirectory}\Modules;$TestToolsModulesPath"
+
+    $startupArgs =  "Set-ExecutionPolicy Bypass -Force -Scope Process; `$env:PSModulePath = '${updatedEnvPath}';"
+    $targetArgs = "${startupArgs}", "Invoke-Pester","${TestDirectory}"
```
Should we depend on Start-PSPester instead as we've started building more specialization in our test runs in that cmdlet?

adityapatwardhan replied:
Start-PSPester launches powershell.exe, we need to launch opencover.console.exe which would launch powershell.exe to collect coverage numbers. Adding this logic to build.psm1 does not seem right and would complicate the implementation on Start-PSPester further.

SteveL-MSFT replied:
For example, start-pspester builds some test exes used by some of the tests.  Perhaps not in this PR, but eventually we should probably consider having a separate module for test execution (perhaps merge Start-PSPester and OpenCover module?)

adityapatwardhan replied:
Yes, I like the idea of have a separate module for test execution. It would have 2 modes, with code coverage and regular.

adityapatwardhan commented:
@SteveL-MSFT Updated. Please have a look.

adityapatwardhan committed:
Addressed CR comments

adityapatwardhan commented:
@SteveL-MSFT Had forgotten to push. Please have a look.

adityapatwardhan commented:
@JamesWTruher  Can you have a look too?

adityapatwardhan committed:
Changed parameter -Quiet to -SuppressQuiet

JamesWTruher commented on test/tools/CodeCoverageAutomation/Start-CodeCoverageRun.ps1:
```diff
@@ -168,22 +169,27 @@ try
 
     $openCoverParams = @{outputlog = $outputLog;
-        TestDirectory = $testPath;
+        TestPath = $testPath;
```
since these hashtable entries are on separate lines, the `;` is not really needed. I don't think it's a big deal, but something to watch in future

adityapatwardhan replied:
Fixed.

JamesWTruher commented on test/tools/CodeCoverageAutomation/Start-CodeCoverageRun.ps1:
```diff
@@ -168,22 +169,27 @@ try
 
+    if($SuppressQuiet)
+    {
+        $openCoverParams.Add('$SuppressQuiet', $true)
```
should `$SuppressQuiet` be `SuppressQuiet`?

adityapatwardhan replied:
Good catch. Fixed.

JamesWTruher commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -524,15 +518,32 @@ function Invoke-OpenCover
+    $base64targetArgs = [convert]::ToBase64String($bytes)
+
+    # the order seems to be important. Always keep -targetargs as the last parameter.
+    $cmdline = "-target:$target","-register:user","-output:${outputLog}","-nodefaultfilters","-oldstyle","-hideskipped:all","-mergeoutput", "-filter:`"+[*]* -[Microsoft.PowerShell.PSReadLine]*`"", "-targetargs:`"-EncodedCommand $base64targetArgs`""
```
having these all on one line is a difficult to read, this would work as well:
```PowerShell
$cmdline = "-target:$target",
    "-register:user",
    "-output:${outputLog}",
    "-nodefaultfilters",
    "-oldstyle",
    "-hideskipped:all",
    "-mergeoutput",
    "-filter:`"+[*]* -[Microsoft.PowerShell.PSReadLine]*`"",
    "-targetargs:`"-EncodedCommand $base64targetArgs`""
```

adityapatwardhan replied:
Fixed.

adityapatwardhan replied:
Also converted it to a single string using -join " " before returning

JamesWTruher commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -476,46 +479,37 @@ function Invoke-OpenCover
+            "$openCoverBin $cmdlineUnelevated" | Out-File -FilePath "$env:temp\unelevated.ps1" -Force
             runas.exe /trustlevel:0x20000 "powershell.exe -file $env:temp\unelevated.ps1"
             # wait for process to start
             Start-Sleep -Seconds 5
```
this 5 second sleep is probably not needed since you have the sleep in line 515

adityapatwardhan replied:
Removed the wait.

JamesWTruher reviewed (approved):
except for the comment about `$SuppressQuiet` this is fine

adityapatwardhan committed:
Addressed CR feedback

adityapatwardhan commented:
@SteveL-MSFT @JamesWTruher I have addressed all feedback, please have a look.

SteveL-MSFT reviewed (approved):
LGTM

daxian-dbw commented on test/tools/CodeCoverageAutomation/Start-CodeCoverageRun.ps1:
```diff
@@ -1,7 +1,8 @@
     [Parameter(Mandatory = $true, Position = 1)] $codecovToken,
-    [Parameter(Position = 2)] $azureLogDrive = "L:\"
+    [Parameter(Position = 2)] $azureLogDrive = "L:\",
+    [Parameter] [switch] $SuppressQuiet
```
This will make the parameter type of `SuppressQuiet` be [parameter], for example:
```
function bar {
param(
  [Parameter] [switch] $SuppressQuiet
)
}

PS> gcm bar -Syntax
bar [[-SuppressQuiet] <Parameter>]
```

adityapatwardhan replied:
Fixed.

adityapatwardhan committed:
Removed Parameter type

adityapatwardhan committed:
Fix unrecognized cmdline parameter error of opencover.console.exe

Opencover.console.exe gets confused when the base64 encoded parameter is given with '&' invoke.
Writing to a ps1 file and invoking the script works around the issue.
This also makes it similar to how unelevated tests are invoked.

adityapatwardhan commented:
@daxian-dbw @SteveL-MSFT  please have another look, I have addressed the feedback and added another bug fix.
